[Room][Compiler Processing] Adds configuration to define processor target language

gfreivasc opened the pull request:
## Proposed Changes

Adds configuration that defines if a processing environment should
expect to work as if on a regular Java setup or if it's Kotlin only.
This defines internal behavior as to how some things will be modelled,
for example if we can expect top level methods to be modelled as such,
or if we should synthesize a `Kt` type to hold it instead.

For JavaAP, the `XProcessingStep.asAutoCommonProcessor()` method now
expects the language to be specified. If not specified, it will default
to expect Java language. Same is applied for KSP.

## Testing
Test:

    ./gradlew \
        test connectedCheck \
        -x :room:room-benchmark:cC \
        -x :room:integration-tests:room-incremental-annotation-processing:test

## Issues Fixed

Fixes: [b/185760236](https://issuetracker.google.com/issues/185760236)

gfreivasc committed:
Adds configuration to define processor target language

Adds configuration that defines if a processing environment should
expect to work as if on a regular Java setup or if it's Kotlin only.
This defines internal behavior as to how some things will be modelled,
for example if we can expect top level methods to be modelled as such,
or if we should synthesize a `Kt` type to hold it instead.

For JavaAP, the `XProcessingStep.asAutoCommonProcessor()` method now
expects the language to be specified. If not specified, it will default
to expect Java language. KSP only expects Kotlin language.

yigit commented:
for the lint failure, you need 

`Test:`

`Bug:` or `Fixes:`

stanzas (we should probably update the template)

Btw, overall lgtm

gfreivasc commented:
I've updated the description after it failed the test but I would need to trigger the pipeline again

yigit commented on room/compiler-processing/src/main/java/androidx/room/compiler/processing/XProcessingEnv.kt:
```diff
@@ -128,11 +134,19 @@ interface XProcessingEnv {
         KSP
     }
 
+    enum class Language {
+        JAVA,
+        KOTLIN
+    }
+
```
sorry i forgot to put comments here but it would be good to add comments here.

```suggestion
    /**
     * The target language for XProcessingEnv.
     * This value controls certain logic when processing code, see the documentation for [XProcessingEnv.targetLanguage] for details.
     */
    enum class Language {
        JAVA,
        KOTLIN
    }

```

gfreivasc replied:
Indeed, done.

yigit commented on room/compiler-processing/src/main/java/androidx/room/compiler/processing/XProcessingEnv.kt:
```diff
@@ -33,6 +33,12 @@ import kotlin.reflect.KClass
+
+    /**
+     * The target language the processor is set to expect
+     */
```
we don't have anything here yet but i think this is probably the best place to document behavior changes between the two. As we add them, we should start documenting them here (open for better suggestions though )

yigit commented on room/compiler-processing/src/main/java/androidx/room/compiler/processing/XProcessingStep.kt:
```diff
@@ -85,7 +88,7 @@ internal class JavacProcessingStepDelegate(
         // create a new x processing environment for each step to ensure it can freely cache
         // whatever it wants and we don't keep elements references across rounds.
-        val xEnv = JavacProcessingEnv(env)
+        val xEnv = JavacProcessingEnv(env, targetLanguage)
```
i'm not sure if we can support kotlin here so we might consider eventually throwing Unsupported exception from the constructor of this but fine for this PR until we clarify what to do there.

gfreivasc replied:
The processor doesn´t work with multiple rounds if it generates kotlin code, but I thought this API was fine.

yigit replied:
i was more thinking in the lines of "can we fake kotlin structure from java processor". e.g. we would need to detect things like "this class is actually generated from a KSFile" etc, which i thought would be hard.

yigit reviewed (approved):
approving to kick of aosp (not all checks run here yet)

gfreivasc committed:
Adds documentation to `XProcessing.Language` type

gfreivasc committed:
Makes KSP target language configurable

By default, KSP will expect to target Java code, this may require some
client code changes in case the processor runs KSP targeting Kotlin code.

yigit commented on room/compiler-processing/src/main/java/androidx/room/compiler/processing/XProcessingEnv.kt:
```diff
@@ -141,12 +160,14 @@ interface XProcessingEnv {
             codeGenerator: CodeGenerator,
-            logger: KSPLogger
+            logger: KSPLogger,
+            targetLanguage: Language = Language.KOTLIN
```
we should have the same default for all and since it works like Java now, this should always be JAVA.

gfreivasc replied:
Yeah, that´s true, changed that part.

yigit commented on room/compiler-processing/src/main/java/androidx/room/compiler/processing/ksp/KspProcessingEnv.kt:
```diff
@@ -38,7 +38,8 @@ internal class KspProcessingEnv(
     logger: KSPLogger,
-    val resolver: Resolver
+    val resolver: Resolver,
+    override val targetLanguage: XProcessingEnv.Language = XProcessingEnv.Language.KOTLIN
```
ditto. this should default to java.

though i'm wondering, given that we require these in the builders, can we remove the default here?

gfreivasc replied:
We could, but I was wondering if this could break client code. Will see if it breaks compilation or tests. If it passes I´ll remove it.

gfreivasc replied:
[`SyntheticKspProcessor`](https://github.com/androidx/androidx/blob/788e5b8a5af3ca6102837b93ed272c60d4945f27/room/compiler-processing-testing/src/main/java/androidx/room/compiler/processing/SyntheticKspProcessor.kt#L54) seems to make use of that factory. Should I explicitly pass Java to it?

yigit replied:
yep. at least for now until we add this language parameter to the `runProcessorTest` function.

yigit replied:
btw you can also implement `runProcessorTest` change here but can be a followup too. whatever you prefer.

gfreivasc replied:
Will take a look into it

yigit commented on room/compiler-processing-testing/src/main/java/androidx/room/compiler/processing/SyntheticKspProcessor.kt:
```diff
@@ -55,7 +55,8 @@ class SyntheticKspProcessor private constructor(
             codeGenerator,
-            logger
+            logger,
+            XProcessingEnv.Language.JAVA
```
we should probably pass this from somewhere into here instead of hardcoding. 

though after looking at the rest of the CL, you've not plugged it yet (to runProcessorTest). Were you planning to do that as a followup? It is probably better together so that there is a way to invoke it and then we can also write a runProcessorTest to ensure it is properly passed around.

gfreivasc replied:
I was thinking of doing it as a follow up, indeed. Will be doing it in this PR.

gfreivasc committed:
Wires target lang configuration to test methods

Changes `runProcessorTest` and similar methods to receive a
targetLanguage parameter which is set to JAVA by default. This is passed
down to the test backend through `TestCompilationParameters` class which
was modified to receive this configuration.

These changes reflected on the construction of `SyntheticKspProcessor`
used in tests which is no longer hardcoded to work with JAVA.

gfreivasc commented:
@yigit I've modified `runProcessorTest` and similar methods to take the target language as a parameter, thus removing the hardcoded value from the synthetic test processor which now takes that information from the constructor instead. Let me know if there are any further changes you think are applicable to these APIs.

Also, I've changed the PR description to be more in line with the fact that Java is the default target language regardless of whether it's running JavaAP or KSP.

gfreivasc commented on room/compiler-processing-testing/src/main/java/androidx/room/compiler/processing/util/runner/JavacCompilationTestRunner.kt:
```diff
@@ -33,7 +33,10 @@ internal object JavacCompilationTestRunner : CompilationTestRunner {
 
     override fun compile(params: TestCompilationParameters): CompilationResult {
-        val syntheticJavacProcessor = SyntheticJavacProcessor(params.handlers)
+        val syntheticJavacProcessor = SyntheticJavacProcessor(
+            params.handlers,
+            params.targetLanguage
+        )
```
I wonder if it wouldn't be some kind of invalid scenario to have Javac targeting kotlin. The `runJavaProcessorTest()` is not supposed to run with kotlin source for what I can tell.

yigit replied:
yea this never runs with kotlin sources (in that case, we use kapt instead of javac).
We can make the processing env throw if language is not set to java though that would make testing a bit harder here :/ i'm fine to leave it as is for now.

gfreivasc replied:
I think I have an idea of how to deal with it here. Will be changing that along with @danysantiago's suggestion.

gfreivasc commented:
@yigit are these changes still relevant/applicable or should we close this PR?

yigit commented:
@gfreivasc sorry i lost track of this.
The changes a still relevant, last time i checked, i had some comments for change but then missed your comments....

Sorry about that, will re-review.

danysantiago commented on room/compiler-processing/src/main/java/androidx/room/compiler/processing/XProcessingEnv.kt:
```diff
@@ -33,6 +33,12 @@ import kotlin.reflect.KClass
     val backend: Backend
+
+    /**
+     * The target language the processor is set to expect
```
nit: Reword a bit to clarify: `The expected language of the sources generated by the processor.`

gfreivasc replied:
Changed the description

gfreivasc committed:
Changes Javac TestRunner target interaction

Explicitly states that JavacCompilationTestRunner is not supposed to
handle KOTLIN as a target language (and Kapt should be used instead).
The API surrounding it has been modified to not take that as a
parameter. If however it's still called to run on top of Kotlin code, it
will throw an error.

gfreivasc commented:
@yigit I made changes to make `JavacCompilationtestRunner` throw if it receives KOTLIN as a target Language, and also changed the code that calls it to avoid passing Kotlin down as a parameter. There are tests to assert these behaviors.
